service/controller/v2/resource/deployment/resource.go

```diff
@@ -97,16 +99,38 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 			},
 		},
 	}
-	f, err := deploymentsClient.CreateOrUpdate(ctx, resourceGroupName, mainDeployment.Name, d)
-	if err != nil {
+
+	d, err := deploymentsClient.Get(ctx, resourceGroupName, mainDeployment.Name)
+	if IsNotFound(err) {
+		_, err := deploymentsClient.CreateOrUpdate(ctx, resourceGroupName, mainDeployment.Name, newDeployment)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deployment in progress")
+		reconciliationcanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation for custom object")
+
+		return nil
+	} else if err != nil {
 		return microerror.Mask(err)
+	} else {
+		s := *d.Properties.ProvisioningState
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", s))
+
+		if !isFinalState(s) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation for resource")
+
+			return nil
+		}
 	}
-	err = f.WaitForCompletion(ctx, deploymentsClient.Client)
+
+	_, err = deploymentsClient.CreateOrUpdate(ctx, resourceGroupName, mainDeployment.Name, newDeployment)
```

Comment:
Should not we have the same logic for `ProvisioningState` here ?

Otherwise we will start reconciling other resources which might still be updating, not sure about the effect of this.

Reply:
When we get here we have a final state and trigger the reconciliation again. This is desired IMO. Uncertain about side effects. I think we should fix problems once we have them. Right now there are no problems I am aware of with this approach. The only thing Pawel mentioned already is that other resources might have problems when cluster creation processes take more time than usual. I will check on this and we already track an issue for this.

Reply:
Cool if we track that somewhere

Reply:
I addressed this here now.